lvmd/command/lvm_command.go

```diff
@@ -0,0 +1,161 @@
+package command
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var Containerized = false
+
+// callLVM calls lvm sub-commands and prints the output to the log.
+func callLVM(ctx context.Context, args ...string) error {
+	return callLVMInto(ctx, nil, args...)
+}
+
+// callLVMInto calls lvm sub-commands and decodes the output via JSON into the provided struct pointer.
+// if the struct pointer is nil, the output will be printed to the log instead.
+func callLVMInto(ctx context.Context, into any, args ...string) error {
+	output, err := callLVMMStreamed(ctx, args...)
+	if err != nil {
+		return fmt.Errorf("failed to execute command: %v", err)
+	}
+
+	// if we don't decode the output into a struct, we can still log the command results from stdout.
+	if into == nil {
+		scanner := bufio.NewScanner(output)
+		for scanner.Scan() {
+			log.FromContext(ctx).Info(strings.TrimSpace(scanner.Text()))
+		}
+		err = scanner.Err()
+	} else {
+		err = json.NewDecoder(output).Decode(&into)
+	}
+	closeErr := output.Close()
```

Comment:
The reason I am not using defer here is because I need to join the errors. This is because during Close, the Wait method gets called and this is where the Process Exit Code Error and also the Stderr output will appear (which was previously not logged at all).

There is only one way I could think of to get rid of the non-deferred close here: to extract the cmd.Wait() out of the io.ReadCloser's Close method and have it called manually. For simplicities sake, I decided to combine these two in the Close call.

Reply:
I thought we could use `defer` here by defining the variable `err` at the very beginning of the function, that is,

```go
func callLVMInto(ctx context.Context, into any, args ...string) (err error) { // <--- define `err` here
	output, err := callLVMMStreamed(ctx, args...)
	if err != nil {
		return fmt.Errorf("failed to execute command: %v", err)
	}
	defer func() { // <-- use `defer` to close `output` and join the errors
		err = errors.Join(output.Close(), err)
	}()

	// if we don't decode the output into a struct, we can still log the command results from stdout.
	if into == nil {
		scanner := bufio.NewScanner(output)
		for scanner.Scan() {
			log.FromContext(ctx).Info(strings.TrimSpace(scanner.Text()))
		}
		err = scanner.Err()
	} else {
		err = json.NewDecoder(output).Decode(&into)
	}
	return // <--- Just return
}
```

cf.: https://wstrm.dev/posts/errors-join-heart-defer/

Does this resolve your concern?

Reply:
While I understand that his works, is this more readable code? If you think it is I am happy to change it accordingly.

Reply:
IMO, Jakob's original code is clearer.

Reply:
@jakobmoellerdev @satoru-takeuchi Thanks for your feedback. Let's keep it that way.